Inferring region layer info and displaying read-only

sdevalapurkar opened the pull request:
# Overview

## Links to jira tickets

- https://quartech.atlassian.net/browse/BHBC-1258
- https://quartech.atlassian.net/browse/BHBC-1260

## This PR contains the following changes

- Implement UI and API changes to support project and survey boundaries
- DB changes for project and surveys to link boundaries to parks, regions, WMU

- Pending to fix UI and make cleaner

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [x] Enhancement (improvements to existing functionality)

## How Has This Been Tested?

- Locally

## Screenshots

![image](https://user-images.githubusercontent.com/28017034/127198196-9e97d3fe-a349-47f8-a601-79b047460fb2.png)

sdevalapurkar committed:
wip

sdevalapurkar committed:
try to fix migration

sdevalapurkar committed:
fix formatting

sdevalapurkar committed:
fix tests

sdevalapurkar committed:
remove park and management unit

cgarrettjones commented on api/src/queries/project/project-delete-queries.ts:
```diff
@@ -3,40 +3,6 @@ import { getLogger } from '../../utils/logger';
-    return null;
-  }
-
-  const sqlStatement: SQLStatement = SQL`
```
just to confirm, the intent is the delete **all** regions associated with a project? if the intent is to only delete a particular region associated with a project then use project_region_id to identify the record to delete.

cgarrettjones replied:
i guess we should get in the habit of checking users association with the data and their roles as well?

NickPhura replied:
This particular change is just showing that shreyas has deleted this whole file.  Since we plan to infer the region based on the boundary, we dont need to store region information anymore, and so the region table can be removed and all associated code, which is a large portion of the changes in this PR.

cgarrettjones commented on database/src/migrations/20210726170005_remove_project_region.ts:
```diff
@@ -0,0 +1,72 @@
+import * as Knex from 'knex';
```
sorry, not sure what this migration file is for... is this not part of the base install?

NickPhura replied:
It looks like Shreyas edited the existing migrations to remove all of the foreign key references to the region table, and then dropped region table itself in this new migration.

NickPhura replied:
But, we could move this change into the existing migration, if you prefer.  So we just edit the main SQL file to not create the region table at all, and then this new migration file could be removed. Normally we'd prefer doing new migrations, but as the changes are in both places, we could unify them in this case.

cgarrettjones reviewed (dismissed):
lots of work there, nice job!

sdevalapurkar committed:
phew

anissa-agahchen reviewed (commented):
For the database changes, use migrations instead of changing the release bundle, because the release bundle is a baseline, and should reflect the latest db release that Charlie put up.

anissa-agahchen committed:
Merge branch 'dev' of https://github.com/bcgov/biohubbc into BHBC-1259

NickPhura committed:
BHBC-1259: Move some code to separate components/functions
- Minor update to error handling
- Minor update to inferred information array to use a Set instead

NickPhura committed:
Merge branch 'BHBC-1259' of https://github.com/bcgov/biohubbc into BHBC-1259

NickPhura committed:
BHBC-1259: git add WFSFeaturePopup.tsx

anissa-agahchen committed:
migrations

anissa-agahchen committed:
Merge branch 'BHBC-1259' of https://github.com/bcgov/biohubbc into BHBC-1259

anissa-agahchen committed:
Merge branch 'dev' into BHBC-1259

anissa-agahchen committed:
clean up migration files

sdevalapurkar committed:
fix multipolygon and point issue

sdevalapurkar committed:
Merge branch 'BHBC-1259' of https://github.com/bcgov/biohubbc into BHBC-1259

sdevalapurkar committed:
fix tests

sdevalapurkar committed:
add missing tests

sdevalapurkar committed:
fix issue with timestamp not coming through

anissa-agahchen commented:
In the edit location/project boundary, editing the region doesn't work.

Meaning:  I cannot delete/add a region after the project is made

sdevalapurkar commented:
> In the edit location/project boundary, editing the region doesn't work.
> 
> Meaning: I cannot delete/add a region after the project is made

cool, please check now with my most recent commit :) thanks!

sdevalapurkar committed:
fix refresh issue

sdevalapurkar committed:
fix issue

sdevalapurkar committed:
fix format and test

NickPhura commented on app/src/components/map/MapContainer.tsx:
```diff
@@ -105,6 +203,32 @@ const MapContainer: React.FC<IMapContainerProps> = (props) => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [preDefinedGeometry]);
 
+  useEffect(() => {
+    if (!geometryState?.geometry.length || !nonEditableGeometries?.length) {
+      setInferredLayersInfo &&
+        setInferredLayersInfo({
+          parks: [],
+          nrm: [],
+          env: [],
+          wmu: []
+        });
+    }
+
+    const isAllPointGeometries = geometryState?.geometry.every((geo: Feature) => geo.geometry.type === 'Point');
+
+    if (isAllPointGeometries) {
+      throttledGetFeatureDetails(layersToInfer);
+    }
+  }, [geometryState?.geometry, nonEditableGeometries]);
+
+  useEffect(() => {
+    if (!bounds || !bounds.length) {
+      return;
+    }
+
+    throttledGetFeatureDetails(layersToInfer);
+  }, [bounds]);
```
Any reason why this wouldn't work? I tried it locally and it seems to work, though I don't know all of the edge cases to test, so maybe you can verify it.

```suggestion
  useEffect(() => {
    if (!geometryState?.geometry.length || !nonEditableGeometries?.length) {
      setInferredLayersInfo &&
        setInferredLayersInfo({
          parks: [],
          nrm: [],
          env: [],
          wmu: []
        });
    }

      throttledGetFeatureDetails(layersToInfer);
  }, [geometryState?.geometry, nonEditableGeometries]);
```

NickPhura replied:
The idea being that we just update the feature details on geometry change, rather than bounds + all points (which seemed like a bit of an odd set of triggers for the side effect we want).

sdevalapurkar replied:
great idea!

sdevalapurkar committed:
feedback

sdevalapurkar committed:
final commit
